ykrt/src/mt.rs

```diff
@@ -34,6 +34,62 @@ const PHASE_COUNTING: usize = 0b100; // Value specifies how many times we've see
 const PHASE_LOCKED: usize = 0b101; // No associated value.
 const PHASE_DONT_TRACE: usize = 0b110; // No associated value.
 
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+struct State(usize);
+
+impl State {
+    fn phase(&self) -> usize {
+        self.0 & PHASE_TAG
+    }
+
+    fn phase_compiled<I>(trace: Box<CompiledTrace<I>>) -> Self {
+        let ptr = Box::into_raw(trace);
+        debug_assert_eq!(ptr as usize & PHASE_TAG, 0);
+        State(ptr as usize | PHASE_COMPILED)
+    }
+
+    fn phase_tracing(tid: Arc<ThreadIdInner>) -> Self {
+        let ptr = Arc::into_raw(tid);
+        debug_assert_eq!(ptr as usize & PHASE_TAG, 0);
+        State(ptr as usize | PHASE_TRACING)
+    }
+
+    fn phase_tracing_lock() -> Self {
+        State(PHASE_TRACING_LOCK)
+    }
+
+    fn phase_compiling<I>(trace: Arc<CompilingTrace<I>>) -> Self {
+        let ptr: *const CompilingTrace<I> = Arc::into_raw(trace);
+        debug_assert_eq!(ptr as usize & PHASE_TAG, 0);
+        State(ptr as usize | PHASE_COMPILING)
+    }
+
+    fn phase_counting(count: usize) -> Self {
+        debug_assert_eq!(count << PHASE_NUM_BITS >> PHASE_NUM_BITS, count);
+        State(PHASE_COUNTING | (count << PHASE_NUM_BITS))
+    }
+
+    fn phase_locked() -> Self {
+        State(PHASE_LOCKED)
+    }
+
+    fn phase_dont_trace() -> Self {
+        State(PHASE_DONT_TRACE)
+    }
+
+    fn number_data(&self) -> usize {
```

Comment:
I think we can add something like `debug_assert_eq!(self.phase(), PHASE_COUNTING)` inside this function.

Reply:
Done